pkg/sql/catalog/descs/descriptor.go

```diff
@@ -163,20 +154,14 @@ func (tc *Collection) getByName(
 	avoidCached, mutable bool,
 ) (found bool, desc catalog.Descriptor, err error) {
 
-	// Handle special specific behaviors for the implied type.
 	var parentID, parentSchemaID descpb.ID
-	switch {
-	case db == nil && sc == nil: // database
-		if db := maybeGetSystemDatabase(name, mutable); db != nil {
-			return true, db, nil
+	if db != nil {
+		if sc == nil {
+			// Schema descriptors are handled in a special way, see getSchemaByName
+			// function declaration for details.
+			return getSchemaByName(ctx, tc, txn, db, name, avoidCached, mutable)
 		}
-	case sc == nil: // schema
-		return getSchemaByName(ctx, tc, txn, db, name, avoidCached, mutable)
-	default: // object
 		parentID, parentSchemaID = db.GetID(), sc.GetID()
-		// Note that we do not attempt to resolve virtual objects here.
-		// We resolve virtual objects by name at a higher level.
-		avoidCached = avoidCached || !canUseLeasingForObject(parentID, name)
```

Comment:
The logic for deciding what can be leased now lives in `leased_descriptor.go`, as it should.